Generate coredumps on signals

mikem8361 opened the pull request:
On MacOS, coredumps are generated if there is an unhandled managed exception
but signals don't generate them.

Enable a stripped down version of the Linux signal handlers for MacOS.

Add a SIGABRT handler so calls to abort() in other native code causes dump to be generated.

Add COMPlus_EnableDumpOnSigTerm that enables dump generation on 
SIGTERM.

Issue: https://github.com/dotnet/runtime/issues/51268

Commit:
Generate coredumps on signals

On MacOS, coredumps are generated if there is an unhandled managed exception
but signals don't generate them.

Enable a stripped down version of the Linux signal handlers for MacOS.

Add a SIGABRT handler so calls to abort() in other native code causes dump to be generated.

Add COMPlus_EnableDumpOnSigTerm that enables dump generation on SIGTERM.

Commit:
Forgot to restore_signal the SIGABRT handler

janvorli commented on src/coreclr/pal/src/exception/signal.cpp:
```diff
@@ -690,6 +723,11 @@ static void sigterm_handler(int code, siginfo_t *siginfo, void *context)
 
         g_pSynchronizationManager->SendTerminationRequestToWorkerThread();
+
+        if (g_enableDumpOnSigTerm)
```
What is the scenario when we want a dump on SIGTERM? It seems problematic, as the dump generation will now run in parallel with the regular SIGTERM handling. The SendTerminationRequestToWorkerThread just pings another thread and returns.

mikem8361 replied:
The point of these changes to get parity with Mono for VSMac when they start running on .NET Core.  And Mono (I'm double checking - see the issue above) generates their crash info json file on SIGTERM. 

You did point out a race that I'll look into.  I'll probably have to move this new code to TerminationRequestHandlingRoutine() in syncmanager.cpp which runs on that termination thread.

janvorli replied:
SIGTERM is currently handled by the runtime by ultimately invoking HandleTerminationRequest in the runtime. That calls ForceEEShutdown(SCA_ExitProcessWhenShutdownComplete); which invokes the Process.Exited event and then shuts down the process. So to avoid disrupting that process, it seems the dump should not be generated before the Process.Exited returns.

jkotas replied:
> The point of these changes to get parity with Mono for VSMac when they start running on .NET Core.

Is configuring this via env variable going to work for VSMac? I see two potential problems: 
- How to set it before the VSMac process starts
- How to clear it for all child processes

mikem8361 replied:
I was definitely going to move it after all the of is called but ForceEEShutdown doesn't return because it ends up calling ExitProcess.  I need to think about the best way to handle this.

mikem8361 replied:
They already have to set the other dump generation (i.e. COMPlus_DbgEnableMiniDump, etc.+)  env vars so I didn't think one more would be a big deal.

mikem8361 replied:
@kdubau how important is generating a dump on SIGTERM? It is more difficult to implement that I thought.

mikem8361 replied:
I think what I just pushed allows SIGTERM to generate coredumps (when  COMPlus_EnableDumpOnSigTerm=1) without the problems pointed out above.

janvorli commented on src/coreclr/pal/src/exception/signal.cpp:
Does SIGSEGV (and other hardware exception signals) handler get ever called? We handle hardware failures using exception ports on macOS and I don't think it composes with signal handling of the same things.

mikem8361 replied:
Yes, they do.  The Mach exception ports get called and then the SIGSEGV, etc. get forwarded through the POSIX signal layer to the signal enable that I enabled.  It is also the only way to catch abort() (SIGABORT) and any signal sent from another process like `kill -s SIGILL <pid>` which isn't that important except for SIGTERM.

Commit:
Moved the "generate dump on SIGTERM" logic out of the PAL into runtime's
HandleTerminationRequest callback and passed a new SCA enum down to
SafeExitProcess that calls TerminateProcess instead of ExitProcess
which generates a core dump.

jkotas commented on src/coreclr/pal/src/thread/process.cpp:
```diff
@@ -3359,6 +3360,8 @@ PROCAbort()
 
     PROCCreateCrashDumpIfEnabled();
 
+    SEHCleanupSignals();
```
Why do we need to call SEHCleanupSignals here? Our policy is to not cleanup on process exit, unless it is absolutely necessary.

mikem8361 replied:
The SIGABRT handler that I added needs to be removed in order to prevent recursion when abort() is called.

jkotas replied:
Is the recursion going to be a problem for other places (e.g. in 3rd party libraries) that may call `abort`?

mikem8361 replied:
It shouldn't because for those abort()s we will catch them, generate the coredump, uninstall the signal handler, and then return from the signal handler in invoke_previous_action in signal.cpp. The normal path through that function doesn't call PROCAbort (fails through to the PROCCreateCrashDumpIfEnabled at the end).  

Even if PROCAbort is called in the invoke_previous_action paths that do, I assume calling abort() again after uninstalling the signal handlers will abort with any problems.

mikem8361 replied:
I've tested this by pinvoking to abort() in my test app.

jkotas replied:
Ok, I would limit the cleanup to just SIGABRT.

KevinCathcart commented on src/coreclr/pal/src/exception/signal.cpp:
```diff
@@ -206,8 +212,10 @@ BOOL SEHInitializeSignals(CorUnix::CPalThread *pthrCurrent, DWORD flags)
+#endif // HAVE_MACH_EXCEPTIONS
     }
 
+#if !HAVE_MACH_EXCEPTIONS
```
Just below this is code setting SIGPIPE to  SIG_IGN.

There already exists code to do this in machexception.cpp. Could we remove that code and un-ifdef the code here, to consolidate all signal handling in this one file? 

https://github.com/dotnet/runtime/blob/d582900145d6ecf752f32b272c6b87c58f4470bd/src/coreclr/pal/src/exception/machexception.cpp#L1419

Commit:
More code review feedback

janvorli reviewed (approved):
LGTM, thank you!
